File: LimSeoyeon/01_Stack/BOJ_1918_후위표기식.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;
int N;
string s;
stack<char> st;
int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> s;
	for (int i = 0; i < s.size(); i++) {
		if (s[i] >= 'A' && s[i] <= 'Z') {
			cout << s[i];
		}
		else if (s[i] == '(') {
			st.push('(');
		}
		else if (s[i] == ')') {
			while (!st.empty()&&st.top()!='(') {
				cout<< st.top();
				st.pop();
			}
			st.pop();
		}
		else { 
			// s[i] 연산자가 우선순위 높은 경우 => pop하지 않고 스택 쌓기
			// 1. st.top이 -,+이고 s[i]가 *,/일때
			// 2. st.top이 (일 때.
			// 3. st가 empty일때

			//s[i] 연산자가 우선순위 낮을 경우 
			// => s[i]가 최고의 우선순위가 될 때까지 pop
			if (st.empty() || st.top() == '(') {
				st.push(s[i]);
			}
			else if (s[i] == '*' || s[i] == '/') {
				while (!st.empty() && st.top()!='(') {
					if (st.top() == '*' || st.top() == '/') {
						cout << st.top();
						st.pop();
					}
					else break;
				}
				st.push(s[i]);
			}
			else if (s[i] == '+' || s[i] == '-') {
				while (!st.empty() && st.top() != '(') {
					cout << st.top();
					st.pop();
				}
				st.push(s[i]);
			}
		}
		
	}
	//다 돌고 나서도 스택에 연산자 남아있을 수 있으므로 LIFO 방식으로 출력.
	while(!st.empty()) {
		cout << st.top();
		st.pop();
	}

}

File: LimSeoyeon/01_Stack/BOJ_1918_후위표기식.java
import java.util.Stack;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner =new Scanner(System.in);
        String str = scanner.next();
        Stack<Character> st = new Stack<>();

        for(int i=0; i<str.length(); i++){
            char ch = str.charAt(i);
            if(ch>='A' && ch<='Z'){
                System.out.print(ch);
            }
            else if(ch=='('){
                st.push(ch);
            }

            else if(ch==')'){
                while (!st.empty()&& st.peek()!='('){
                    System.out.print(st.pop());
                }
                st.pop();
            }
            else{
                if(st.empty()||st.peek()=='('){
                    st.push(ch);
                }
                else if(ch=='*' || ch=='/'){
                    while(!st.empty()&&st.peek()!='('){
                        if(st.peek()=='*' || st.peek()=='/'){
                            System.out.print(st.pop());
                        }
                        else break;
                    }
                    st.push(ch);
                }
                else{ // ch 가 + or -
                    while(!st.empty()&&st.peek()!='('){
                        System.out.print(st.pop());
                    }
                    st.push(ch);
                }
            }

        }

        while(!st.empty()){
            System.out.print(st.pop());
        }

    }

}
